### ISSUE TITLE
[Ray Tune] Set a different experiment folder for storing the experiments locally on a single machine with pytorch Lightning.

### ISSUE THREAD

**Issue by LeWentz** [ray-project/ray#37075]
### What happened + What you expected to happen

**Goal: Set a different experiment folder for storing the experiments locally using Windows 11 on a single machine with pytorch Lightning.**

**Bug**
Setting `local_dir` or `storage_path` parameter in RunConfig does not ensure that only the given path is used to store the experiments and without errors. Instead I got different behaviors, trying different way around it, see reproduction script section below.

**Expected Behavior:**
Setting the new `storage_path` parameter to a local windows file ensures all tune experiments are stored under that folder only without errors.

**Useful Information:**

This seem to not have been an issue with ray 2.4. On my older project I used the `local_dir` parameter of the `RunConfig` which worked just fine. After upgrading to 2.5 and also upgrading pytorch lightning to 2.0.4, the first error came up and required me to install pyarrow and right after that to set "gloo" as backend in the `TorchConfig`. [For me it is not obvious, why that is needed, but I found the solution already in the issues (https://github.com/ray-project/ray_lightning/issues/13). I don't know if it is the right place, but maybe you can consider adding these details to your documentation for windows users.]

I feel like, the issue is caused when handling the folder scheme:
https://github.com/ray-project/ray/blob/4b7ebaabbaa84bdf620247610c28052a136d04c1/python/ray/tune/utils/util.py#L177-L189
in combination with:
https://github.com/python/cpython/issues/86381#issuecomment-1093889674

> Passing a Windows path to urllib.parse.urlparse, without file://, makes it detect the drive as the scheme even it having '\' after the ':', which makes it completelly wrong for a url.

Trying to set an absolut path like: "c:/users/ray/experiments/" results in "c" being the URI scheme, which in turn leads ray to think of this as a remote path.

===

On a sidenote: it remains unclear to me, how `LightningConfigBuilder.trainer.default_root_dir` works in combination with the `storage_path` parameter. The same question arises for `LightningConfigBuilder.checkpoiting()` and the `CheckpointConfig` of the RunConfig. I kindly appreciate, if you can shed a little light on this.

Kind Regards and thank your work!

### Versions / Dependencies

```
Windows 11 22621.1848

python 3.10.9

lightning                    2.0.4
lightning-bolts              0.6.0.post1
lightning-cloud              0.5.37
lightning-utilities          0.8.0
pyarrow                      12.0.1
ray                          2.5.1
slicer                       0.0.7
torch                        2.0.1+cu118
torchaudio                   2.0.2+cu118
torchvision                  0.15.2+cu118
```

### Reproduction script

I was basically following the official documentation: https://docs.ray.io/en/latest/tune/examples/tune-pytorch-lightning.html 
The snippet below shows my configuration.
I tried multiple ways to adjust the experiments folder and got different results. Sadly I cannot reproduce all of them, but so far I remember the following to look out for:

**Test 1** - setting `storage_path` instead of `local_dir`:
The experiments were logged in the default folder (C:\Users\<user>\ray_results\) as well as `self.log_dir`. Both folders basically contained the same content. Maybe `default_root_dir`was the issue here, setting that and the enviromental variable ensures only one folder, but still with an error, causing further code to break. After the experiment was done, it seemed ray was trying to copy file to a remote folder (I shortened the paths for readbility):

> 
>   File "C:\Users\\ray\tune\tuner.py", line 347, in fit
>     return self._local_tuner.fit()
>   File "C:\Users\\ray\tune\impl\tuner_internal.py", line 588, in fit
>     analysis = self._fit_internal(trainable, param_space)
>   File "C:\Users\\ray\tune\impl\tuner_internal.py", line 712, in _fit_internal
>     analysis = run(
>   File "C:\Users\\ray\tune\tune.py", line 1070, in run
>     runner.step()
>   File "C:\Users\\ray\tune\execution\tune_controller.py", line 272, in step
>     raise e
>   File "C:\Users\\ray\tune\execution\tune_controller.py", line 269, in step
>     self.checkpoint()
>   File "C:\Users\\ray\tune\execution\trial_runner.py", line 491, in checkpoint
>     self._checkpoint_manager.checkpoint(
>   File "C:\Users\\ray\tune\execution\experiment_state.py", line 235, in checkpoint
>     self.sync_up(force=force, wait=wait)
>   File "C:\Users\\ray\tune\execution\experiment_state.py", line 265, in sync_up
>     self._syncer.wait()
>   File "C:\Users\\ray\tune\syncer.py", line 616, in wait
>     raise TuneError(f"Sync process failed: {e}") from e
> ray.tune.error.TuneError: Sync process failed: GetFileInfo() yielded path 'C:/Experiments/2DShapes/LightningTrainer_2023-07-04_14-34-40/0_2e393_00000_batch=None', which is outside base dir 'C:\Experiments\2DShapes\LightningTrainer_2023-07-04_14-34-40'

**Test 2** - setting the env "RAY_AIR_LOCAL_CACHE_DIR" and do not set either local_dir nor storage_path:
Basically the same as above.

**Test 3** - trying to use the old `local_dir` and disable the `syncher` by setting it to `None`:
Error `upload_dir` enables syncing to cloud storage, but `syncer=None` disables syncing. Either remove the `upload_dir`, or set `syncer` to 'auto' or a custom syncer.
Yet `upload_dir` is automatically set due to the way the scheme of `local_dir` is handled.

```
lightning_config = (
    LightningConfigBuilder()
    .module(cls=self.tunable_model, config=self.param_space)
    .trainer(max_epochs=self.num_epochs, accelerator='gpu', deterministic=True, default_root_dir=self.log_dir, logger=logger, enable_checkpointing=False)
    .fit_params(datamodule=self.datamodule)
    .checkpointing(monitor=metric, save_top_k=2, mode=self.mode)
    .build()
)

run_config = RunConfig(
    checkpoint_config=CheckpointConfig(
        num_to_keep=3, 
        checkpoint_score_attribute=metric,
        checkpoint_score_order=self.mode,
    ),
    verbose=3,
    log_to_file='std_combined.log',
    progress_reporter=reporter,
    #storage_path=self.log_dir,
    local_dir=self.log_dir,
)

run_config.sync_config.syncer = None # seems also not to work
run_config.sync_config.upload_dir = None

scaling_config = ScalingConfig(
    num_workers=1, use_gpu=True, resources_per_worker={'CPU': self.resources_cpu, 'GPU': self.resources_gpu}, 
)

lightning_trainer = LightningTrainer(
    lightning_config=lightning_config,
    scaling_config=scaling_config,
    run_config=run_config,
    torch_config=TorchConfig(backend='gloo'),
)

tuner = tune.Tuner(
    lightning_trainer,
    param_space={'lightning_config': lightning_config},
    tune_config=tune.TuneConfig(
        metric=metric, 
        mode=self.mode,
        num_samples=self.num_samples,
    ),
)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by 2505928188** [comment]
it seems revising “if parsed.scheme:” to "if False: " in line 185 could fix this bug  temporarily.

**Comment by justinvyu** [comment]
This is fixed from Ray 2.7 onward. See https://docs.ray.io/en/latest/train/user-guides/persistent-storage.html
